instrumentation/spring/spring-boot-autoconfigure/src/main/java/io/opentelemetry/instrumentation/spring/autoconfigure/internal/properties/CachedPropertyResolver.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.spring.autoconfigure.internal.properties;
+
+import java.util.Objects;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import javax.annotation.Nullable;
+import org.springframework.core.env.Environment;
+
+/**
+ * A caching wrapper around Spring's {@link Environment} that caches property lookups to avoid
+ * repeated expensive operations and property source traversal.
+ *
+ * <p>Thread-safe for concurrent access. Cached values persist indefinitely and are assumed to be
+ * immutable after the first access. If properties can change at runtime, use {@link #clear()} to
+ * invalidate the cache.
+ *
+ * <p>This class is internal and is hence not for public use. Its APIs are unstable and can change
+ * at any time.
+ */
+final class CachedPropertyResolver {
+  private final Environment environment;
+  private final ConcurrentHashMap<CacheKey, Optional<?>> cache = new ConcurrentHashMap<>();
+
+  CachedPropertyResolver(Environment environment) {
+    this.environment = Objects.requireNonNull(environment, "environment");
+  }
+
+  /**
+   * Gets a property value from the environment, using a cache to avoid repeated lookups.
+   *
+   * @param name the property name
+   * @param targetType the target type to convert to
+   * @return the property value, or null if not found
+   */
+  @Nullable
+  <T> T getProperty(String name, Class<T> targetType) {
+    CacheKey key = new CacheKey(name, targetType);
+    // CacheKey includes targetType, ensuring type match
+    @SuppressWarnings("unchecked")
+    Optional<T> result =
+        (Optional<T>)
+            cache.computeIfAbsent(
+                key, k -> Optional.ofNullable(environment.getProperty(name, targetType)));
+    return result.orElse(null);
```

Comment:
What does using `Optional` give over just using 
`return targetType.cast(cache.computeIfAbsent(key, k -> environment.getProperty(name, targetType)));`

Reply:
ConcurrentHashMap doesn't store `null` values, so if the lookup returns null it'll keep checking the environment instead of caching it